agent/structs/config_entry_discoverychain.go

```diff
@@ -0,0 +1,746 @@
+// protocol as http-based (in centralized configuration).
+type ServiceRouterConfigEntry struct {
+	Kind string
+	Name string
+
+	// Routes is the list of routes to consider when processing L7 requests.
+	// The first rule to match in the list is terminal and stops further
+	// evaluation.
+	//
+	// Traffic that fails to match any of the provided routes will be routed to
+	// the default service.
+	Routes []ServiceRoute
+
+	RaftIndex
+}
+
+func (e *ServiceRouterConfigEntry) GetKind() string {
+	return ServiceRouter
+}
+
+func (e *ServiceRouterConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+
+	return e.Name
+}
+
+func (e *ServiceRouterConfigEntry) Normalize() error {
+	if e == nil {
+		return fmt.Errorf("config entry is nil")
+	}
+
+	e.Kind = ServiceRouter
+
+	// TODO(rb): anything to normalize?
+
+	return nil
+}
+
+func (e *ServiceRouterConfigEntry) Validate() error {
+	if e.Name == "" {
+		return fmt.Errorf("Name is required")
+	}
+
+	// TODO(rb): enforce corresponding service has protocol=http
+
+	// TODO(rb): validate the entire compiled chain? how?
+
+	// TODO(rb): validate more
+
+	// Technically you can have no explicit routes at all where just the
+	// catch-all is configured for you, but at that point maybe you should just
+	// delete it so it will default?
+
+	return nil
+}
+
+func (e *ServiceRouterConfigEntry) CanRead(rule acl.Authorizer) bool {
+	return canReadDiscoveryChain(e, rule)
+}
+
+func (e *ServiceRouterConfigEntry) CanWrite(rule acl.Authorizer) bool {
+	return canWriteDiscoveryChain(e, rule)
+}
+
+func (e *ServiceRouterConfigEntry) GetRaftIndex() *RaftIndex {
+	if e == nil {
+		return &RaftIndex{}
+	}
+
+	return &e.RaftIndex
+}
+
+func (e *ServiceRouterConfigEntry) ListRelatedServices() []string {
+	found := make(map[string]struct{})
+
+	// We always inject a default catch-all route to the same service as the router.
+	found[e.Name] = struct{}{}
+
+	for _, route := range e.Routes {
+		if route.Destination != nil && route.Destination.Service != "" {
+			found[route.Destination.Service] = struct{}{}
+		}
+	}
+
+	out := make([]string, 0, len(found))
+	for svc, _ := range found {
+		out = append(out, svc)
+	}
+	sort.Strings(out)
+	return out
+}
+
+// ServiceRoute is a single routing rule that routes traffic to the destination
+// when the match criteria applies.
+type ServiceRoute struct {
+	Match       *ServiceRouteMatch       `json:",omitempty"`
+	Destination *ServiceRouteDestination `json:",omitempty"`
+}
+
+// ServiceRouteMatch is a set of criteria that can match incoming L7 requests.
+type ServiceRouteMatch struct {
+	HTTP *ServiceRouteHTTPMatch `json:",omitempty"`
+
+	// If we have non-http match criteria for other protocols in the future
+	// (gRPC, redis, etc) they can go here.
+}
+
+// ServiceRouteHTTPMatch is a set of http-specific match criteria.
+type ServiceRouteHTTPMatch struct {
+	PathExact  string `json:",omitempty"`
+	PathPrefix string `json:",omitempty"`
+	PathRegex  string `json:",omitempty"`
```

Comment:
Do we need `path_prefix -> PathPrefix` translations for all of these? We do this for registrations for example but not for other API endpoints. I'm in two minds. We could just not do that at the API/RPC layer and do it in the CLI and config parsing (for bootstrap) instead (maybe we do that part already?)

Reply:
I have not considered snake-to-camel conversions. Is the best example of how it should be done the service registration via a config file (in hcl).

Should the snake-to-camel conversion only apply to HCL forms of these or also json?